lib/dpl/providers/cloudformation.rb

```diff
@@ -0,0 +1,277 @@
+          stream_events(stack_name, :stack_update_complete) if wait?
+          info :done
+        end
+
+        def create
+          promote? ? create_stack : create_change_set(:create)
+        end
+
+        def create_stack
+          info :create_stack
+          params = { timeout_in_minutes: create_timeout, on_failure: 'ROLLBACK' }
+          client.create_stack(common_params.merge(params))
+          stream_events(stack_name, :stack_create_complete) if wait?
+          info :done
+        end
+
+        def create_change_set(type)
+          info :create_change_set
+          set = client.create_change_set(common_params.merge(change_set_params(type)))
+          wait_for(:change_set_create_complete, change_set_name: set.id) if wait? && !test?
+          info :done
+        rescue Aws::Waiters::Errors::FailureStateError => e
+          raise e unless change_set_contains_changes?(set)
+          info :delete_change_set
+          client.delete_change_set(change_set_name: set.id)
+        end
+
+        def change_set_params(type)
+          {
+            change_set_type: type.to_s.upcase,
+            change_set_name: interpolate(str(:change_set_name)),
+            description: interpolate(str(:change_set_desc))
+          }
+        end
+
+        def change_set_contains_changes?(change_set)
+          data = client.describe_change_set(change_set_name: change_set.id)
+          data.status_reason.start_with?(%(The submitted information didn't contain changes))
+        end
+
+        def stack_exists?
+          stack = last_stack
+          stack && stack.stack_status != 'REVIEW_IN_PROGRESS'
+        rescue Aws::CloudFormation::Errors::ValidationError => e
+          raise e unless e.message.include?('does not exist')
+          false
+        end
+
+        def stream_events(stack_name, condition)
+          stream = EventStream.new(client, stack_name, method(:info))
+          wait_for(condition, stack_name: stack_name) unless test? # hmm.
+        ensure
+          stream.stop
+        end
+
+        def wait_for(cond, params)
+          started_at = Time.now
+          timeout = lambda { |*| throw :failure if Time.now - started_at > wait_timeout }
+          # params = params.merge(max_attempts: nil, delay: 5, before_wait: timeout)
+          client.wait_until(cond, params) { |w| w.before_wait(&timeout) }
+        end
+
+        def store_events
+          logs = last_stack.outputs || {}
+          logs = logs.map { |log| "#{log[:output_key]}=#{log[:output_value]}" }
+          File.write(output_file, logs.join("\n"))
+        end
+
+        def last_stack
+          client.describe_stacks(stack_name: stack_name)[:stacks].first
+        end
+
+        def common_params
+          @common_params ||= compact(
+            stack_name: stack_name,
+            role_arn: role_arn,
+            capabilities: capabilities,
+            parameters: parameters,
+            # template: template
```

Comment:
I guess this should be uncommented

Reply:
yes, absolutely :) however, i couldn't get the unit tests accept this. i believe it needs to be `template_body` vs `template_url`, but i'll still need to put more work into this.

@radekl i now have a lot of ports of a lot of PRs open (ported 26 PRs so far), so it might take a little while for me to complete all this. if you would like to take this one back/over, feel free to do so. i'd also happily add you to our Slack dpl community channel for more instant chat. let me know!

Reply:
Ok I will take a look on it, fork it and try to make it work. As for slack dpl community channel - feel free to add me. You can find my email in profile